src/Microsoft.AspNet.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -251,69 +226,134 @@ protected virtual string GenerateUniqueId()
                                 .WriteEndMethodInvocation(endLine: false)   // End of method call
                                 .WriteEndMethodInvocation(endLine: true);   // End of new expression / throw statement
                         }
                     }
-
-                    var dictionaryKey = attributeName.Substring(attributeDescriptor.Name.Length);
-                    valueAccessor = string.Format(
-                        CultureInfo.InvariantCulture,
-                        "{0}.{1}[\"{2}\"]",
-                        tagHelperVariableName,
-                        attributeDescriptor.PropertyName,
-                        dictionaryKey);
                 }
+            }
+        }
+
+        private void RenderAttributes(
+            IList<KeyValuePair<string, Chunk>> chunkAttributes,
+            IEnumerable<TagHelperDescriptor> tagHelperDescriptors)
+        {
+            // First attribute wins, even if there are duplicates.
+            var renderedBoundAttributeNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
 
-                // If we haven't recorded this attribute value before then we need to record its value.
-                var attributeValueRecorded = htmlAttributeValues.ContainsKey(attributeName);
-                if (!attributeValueRecorded)
+            foreach (var attribute in chunkAttributes)
+            {
+                var associatedDescriptors = tagHelperDescriptors.Where(descriptor =>
+                    descriptor.Attributes.Any(attributeDescriptor => attributeDescriptor.IsNameMatch(attribute.Key)));
+
+                // Bound attributes will have associated descriptors.
+                if (associatedDescriptors.Any())
                 {
-                    // We only need to create attribute values once per HTML element (not once per tag helper).
-                    // We're saving the value accessor so we can retrieve it later if there are more tag
-                    // helpers that need the value.
-                    htmlAttributeValues.Add(attributeName, valueAccessor);
-
-                    // Bufferable attributes are attributes that can have Razor code inside of them. Such
-                    // attributes have string values and may be calculated using a temporary TextWriter or other
-                    // buffer.
-                    var bufferableAttribute = attributeDescriptor.IsStringProperty;
-
-                    RenderNewAttributeValueAssignment(
-                        attributeDescriptor,
-                        bufferableAttribute,
-                        attributeValueChunk,
-                        valueAccessor);
-
-                    // Execution contexts are a runtime feature.
-                    if (_designTimeMode)
+                    if (!renderedBoundAttributeNames.Add(attribute.Key))
                     {
+                        // We already rendered an attribute with the same name that was bound. First wins, skip.
```

Comment:
Don't rewrite existing, understandable comments and don't talk in terms of "we".